File: python/day03.py
def count_zeros_ones(l):
    return l.count("0"), l.count("1")


def transpose(arr):
    lines = [list(line) for line in arr]
    return list(map(list, zip(*lines)))


def get_power_rate(lines):
    transposed_lines = transpose(lines)
    epsilon = 0
    gamma = 0
    for line in transposed_lines:
        gamma = gamma << 1
        epsilon = epsilon << 1
        z, o = count_zeros_ones(line)
        if o > z:  # one is most common
            gamma += 1
        elif z > o:  # zero is most common
            epsilon += 1
        else:
            raise Exception("Something went wrong here..")
    return epsilon * gamma


def find_value(lines, criteria):
    max_iterations = len(lines[1])
    mask = ""
    for i in range(0, max_iterations):
        transposed_lines = transpose(lines)
        z, o = count_zeros_ones(transposed_lines[i])
        mask += str(int(not ((z <= o) ^ bool(criteria))))
        lines = list(filter(lambda l: l.startswith(mask), lines))
        if len(lines) == 1:
            break
    return int(lines[0], 2)


report = open("assets/day03.txt").read().splitlines()

print(f"Part A: {get_power_rate(report)}")

print(f"Part B: {find_value(report, 1) * find_value(report, 0)}")

File: src/day03.rs
use transpose;

static INPUT: &str = include_str!("../assets/day03.txt");

fn has_more_ones(l: Vec<char>) -> bool {
    l.iter().filter(|&c| *c == '1').count() >= l.iter().filter(|&c| *c == '0').count() 
}

fn get_power_rate(report: &str) -> u32 {
    let width: usize = report.lines().count();
    let height: usize = report.lines().next().unwrap().len();
    let inp: Vec<char> = report
        .lines()
        .flat_map(|s| s.chars().collect::<Vec<_>>())
        .collect();

    let mut out: Vec<char> = vec!['0'; inp.len()];
    transpose::transpose(&inp, &mut out, height, width);

    let mut gamma: u32 = 0;
    let cs = out.chunks(width);
    for line in cs {
        gamma = gamma << 1;
        gamma += has_more_ones(line.to_vec()) as u32;
    }
    let mask = (2 as u32).pow((height) as u32) -1;
    let epsilon = gamma ^ mask;


    return gamma * epsilon
}


pub fn part_a() -> String {
    get_power_rate(INPUT).to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    static EXAMPLE_INPUT: &str = include_str!("../assets/day03-sample.txt");

    #[test]
    fn example_a() {
        assert_eq!(get_power_rate(EXAMPLE_INPUT), 198);
    }
}
